source/grid/grid_tools.cc

```diff
@@ -3257,24 +3289,31 @@ namespace GridTools
       for (unsigned int f=0; f<GeometryInfo<dim>::faces_per_cell; ++f)
         if (cell->face(f)->at_boundary())
           {
-            // copy boundary to manifold ids
-            cell->face(f)->set_manifold_id
-            (static_cast<types::manifold_id>(cell->face(f)->boundary_id()));
+            const auto bid = cell->face(f)->boundary_id();
+            const auto ind = std::find(src_boundary_ids.begin(), src_boundary_ids.end(), bid)-
+                             src_boundary_ids.begin();
 
-            // then reset boundary ids if so desired, and in 3d also that
-            // of edges
-            if (reset_boundary_ids == true)
+            if ((unsigned int)ind < src_boundary_ids.size())
               {
-                cell->face(f)->set_boundary_id(0);
-                if (dim >= 3)
-                  for (signed int e=0; e<static_cast<signed int>(GeometryInfo<dim>::lines_per_face); ++e)
-                    cell->face(f)->line(e)->set_boundary_id(0);
+                // assign the manifold id
+                cell->face(f)->set_manifold_id(dst_manifold_ids[ind]);
+                // then reset boundary id
+                cell->face(f)->set_boundary_id(reset_boundary_ids[ind]);
               }
+
+            if (dim >= 3)
+              for (unsigned int e=0; e<GeometryInfo<dim>::lines_per_face; ++e)
+                {
+                  const auto bid = cell->face(f)->line(e)->boundary_id();
+                  const auto ind = std::find(src_boundary_ids.begin(), src_boundary_ids.end(), bid)-
+                                   src_boundary_ids.begin();
+                  if ((unsigned int)ind < src_boundary_ids.size())
+                    cell->face(f)->line(e)->set_boundary_id(reset_boundary_ids[ind]);
```

Comment:
@luca-heltai Do we also perhaps need to set the edge manifold id here? Its just struck me now that the call to `cell->face(f)->set_manifold_id()` [does not propagate](https://dealii.org/8.5.0/doxygen/deal.II/group__manifold.html#gac23367ec88e9282815b9a03df8af5431) this information to the edges.

Reply:
Well, I've convinced the setting of edge manifold id's works as expected, so my question can be ignored. I hope you don't mind that I added my test specifically for this purpose.

Reply:
:)

Reply:
btw, instead of doing it by hand for edges, you can call `set_all_boundary_ids()`.

Reply:
Nope. You need to make sure that the original id was the one you want to replace. If you use `set_all_boundary_ids`, this  is done regardless of the original id.